File: quickbeam/pipelines/gem.py
"""Derived rating signals computed purely from a place's own (rating, review-count)
pair — no corpus pass, no user input, no network. Generalizes to anywhere on Earth
and ships inside the static snapshot.

Two facts a general LLM can't give you, because they're relative to a place's own
attention rather than the open web's average:

  - hidden-gem      : genuinely loved but under-discovered (high rating, few reviews)
  - crowd-favorite  : loved and proven at volume (high rating, many reviews)

The hidden-gem signal is, in effect, "the local secret" — surfaced from two numbers
instead of from community notes. `gemScore` is a continuous 0..1 sort key
(quality x obscurity); `ratingTier` is the badge. Both are omitted when there is no
rating to stand on.
"""
import math

# A modest prior the raw rating is shrunk toward, so a 5.0 from 3 reviews doesn't
# masquerade as quality. _K is the pseudo-count ("reviews of prior belief").
_PRIOR, _K = 3.8, 12

# Review count at which a place reads as fully mainstream *anywhere* — rural lake
# town or dense city alike. log-scaled so the obscurity signal is density-agnostic,
# which is what lets the same thresholds generalize beyond any one region.
_MAINSTREAM = 1500
_CAP = math.log10(_MAINSTREAM)


def rating_signal(rating, n) -> dict:
    """Return {"gemScore", optional "ratingTier"} for a (rating, review-count) pair,
    or {} when either is missing. Pure function — safe to call per entity at
    schema-gen time, no corpus required."""
    if not rating or not n:
        return {}

    # Bayesian-shrunk quality in [0,1]: a gem must be genuinely good after shrink,
    # so tiny-n outliers (5.0 from a handful of reviews) don't qualify on rating alone.
    adjusted = (n * rating + _K * _PRIOR) / (n + _K)
    quality = max(0.0, min(1.0, adjusted - 4.0))

    # Obscurity decays with attention; ~_MAINSTREAM reviews -> 0.
    obscurity = max(0.0, 1.0 - math.log10(max(n, 1)) / _CAP)
    score = round(quality * obscurity, 3)

    if adjusted >= 4.5 and obscurity >= 0.30 and score >= 0.20:
        tier = "hidden-gem"
    elif adjusted >= 4.3 and obscurity <= 0.18:
        tier = "crowd-favorite"
    else:
        tier = None

    out = {"gemScore": score}
    if tier:
        out["ratingTier"] = tier
    return out

File: examples/src/lib/gem.ts
// Client mirror of quickbeam/pipelines/gem.py — derives the "hidden gem / local
// favorite" signal from a place's own (rating, review-count) pair. Pure and
// corpus-free, identical to the baked formula, so the badge & sort light up on the
// CURRENT static snapshot with no re-bake (rating + userRatingCount are already in
// the shard). Once the pipeline bakes `ratingTier`/`gemScore` in, those win — see
// ratingSignalOf. Keep the constants in sync with gem.py if either side changes.

export type RatingTier = 'hidden-gem' | 'crowd-favorite';

const PRIOR = 3.8; // modest prior the raw rating is shrunk toward
const K = 12; // pseudo-count ("reviews of prior belief")
const MAINSTREAM = 1500; // review count at which a place reads as fully mainstream anywhere
const CAP = Math.log10(MAINSTREAM);

export interface RatingSignal {
  gemScore: number; // 0..1 continuous sort key (quality x obscurity)
  tier: RatingTier | null; // the badge, when it clears a bar
}

export function ratingSignal(rating?: number | null, n?: number | null): RatingSignal | null {
  if (!rating || !n) return null;
  // Bayesian-shrunk quality in [0,1] — tiny-n 5.0s don't qualify on rating alone.
  const adjusted = (n * rating + K * PRIOR) / (n + K);
  const quality = Math.max(0, Math.min(1, adjusted - 4));
  // Obscurity decays with attention; ~MAINSTREAM reviews -> 0 (density-agnostic).
  const obscurity = Math.max(0, 1 - Math.log10(Math.max(n, 1)) / CAP);
  const gemScore = Math.round(quality * obscurity * 1000) / 1000;

  let tier: RatingTier | null = null;
  if (adjusted >= 4.5 && obscurity >= 0.3 && gemScore >= 0.2) tier = 'hidden-gem';
  else if (adjusted >= 4.3 && obscurity <= 0.18) tier = 'crowd-favorite';
  return { gemScore, tier };
}

// Prefer the baked fields when a re-bake has added them; otherwise derive client-side.
export function ratingSignalOf(fields: Record<string, unknown>): RatingSignal | null {
  const bakedTier = fields.ratingTier;
  const bakedScore = fields.gemScore;
  if (typeof bakedTier === 'string' || typeof bakedScore === 'number') {
    return {
      gemScore: typeof bakedScore === 'number' ? bakedScore : 0,
      tier: bakedTier === 'hidden-gem' || bakedTier === 'crowd-favorite' ? bakedTier : null,
    };
  }
  const rating = typeof fields.rating === 'number' ? fields.rating : null;
  const n = typeof fields.userRatingCount === 'number' ? fields.userRatingCount : null;
  return ratingSignal(rating, n);
}

export const TIER_LABEL: Record<RatingTier, string> = {
  'hidden-gem': 'Hidden gem',
  'crowd-favorite': 'Local favorite',
};
